Forge.TreeWalker/src/TreeWalkerSession.cs

```diff
@@ -884,47 +997,109 @@ public async Task<object> EvaluateDynamicProperty(dynamic schemaObj, Type knownT
         private async Task CommitActionResponse(string treeActionKey, ActionResponse actionResponse)
         {
-            await this.Parameters.ForgeState.Set<ActionResponse>(treeActionKey + ActionResponseSuffix, actionResponse).ConfigureAwait(false);
-            await this.CommitLastTreeAction(treeActionKey).ConfigureAwait(false);
+            List<KeyValuePair<string, object>> itemsToPersist = new List<KeyValuePair<string, object>>();
+
+            itemsToPersist.Add(new KeyValuePair<string, object>(treeActionKey + ActionResponseSuffix, actionResponse));
+            itemsToPersist.Add(new KeyValuePair<string, object>(LastTreeActionSuffix, treeActionKey));
+
+            await this.Parameters.ForgeState.SetRange(itemsToPersist);
         }
 
         /// <summary>
         /// Commits the current tree node to the forgeState.
-        /// The wrapper class has access to this state, allowing it to rehydrate/retry on failures if desired.
+        /// The caller has access to this state, allowing it to rehydrate/retry on failures if desired.
+        /// If we are revisiting a node that has been previously completed and we aren't rehydrating, clear the ActionResponses and Intermediates, and persist PreviousActionResponses.
         /// </summary>
-        /// <param name="treeNodeKey">The TreeNode's key that tree walker is currently walking.</param>
-        private Task CommitCurrentTreeNode(string treeNodeKey)
+        /// <param name="treeNodeKey">The TreeNode's key that tree walker is about to visit.</param>
+        /// <param name="treeNode">The TreeNode that tree walker is about to visit.</param>
+        private async Task CommitCurrentTreeNode(string treeNodeKey, TreeNode treeNode)
         {
-            // TODO: Consider adding a "nodeStep" to save if we already completed the BeforeVisitNode step and should skip straight to VisitNode.
-            return this.Parameters.ForgeState.Set<string>(CurrentTreeNodeSuffix, treeNodeKey);
+            List<KeyValuePair<string, object>> itemsToPersist = new List<KeyValuePair<string, object>>();
+
+            // Handle revisit/cycle case if this node has been previously completed and we aren't rehydrating.
+            if (treeNode.Actions != null && this.hasSessionRehydrated)
+            {
+                // Check if all actions already have an action response.
+                foreach (KeyValuePair<string, TreeAction> kvp in treeNode.Actions)
+                {
+                    string treeActionKey = kvp.Key;
+                    TreeAction treeAction = kvp.Value;
+                    ActionResponse actionResponse = await this.GetOutputAsync(treeActionKey).ConfigureAwait(false);
+
+                    if (actionResponse == null)
+                    {
+                        // If this session has rehydrated and actionResponse is null, then this must be the first time visiting this TreeNode.
+                        itemsToPersist.Clear();
```

Comment:
>itemsToPersist.Clear(); [](start = 24, length = 23)

if an actionNode has 2 actions; one has response saved, and the other one has null for response. Shall we clear for all?

And in this case, the value of itemsToPersist will be decided by which action to check further. If the has-responsed is checked first, then itemsToPersist will be empty. If the has-response is checked secondly, then itemsToPersist has the result of the has-responsed.

Reply:
This part is slightly tricky because it isn't intuitive why I can make this statement:
"If this session has rehydrated and actionResponse is null, then this must be the first time visiting this TreeNode"

But it turns out to be true! If I'm already rehydrated, then to get to this line of code, I have successfully visited a previous node and have childSelected to this node. So there are two possible states for this node:
1. All ActionResponses are persisted and we are revisiting this node.
2. All ActionResponses must be null, and we are visiting this node for the first time.

I try to leave enough comments around to describe this behavior, but let me know if I should add more.

---
In reply to: [356750599](https://github.com/microsoft/Forge/pull/6#discussion_r356750599) [](ancestors = 356750599)